src/rapid_pro_to_engagement_db/rapid_pro_to_engagement_db.py

```diff
@@ -0,0 +1,155 @@
+    Synchronises runs from a Rapid Pro workspace to an engagement database.
+
+    :param google_cloud_credentials_file_path: Path to a Google Cloud service account credentials file to use to access
+                                               the credentials bucket.
+    :type google_cloud_credentials_file_path: str
+    :param rapid_pro_to_engagement_db_configuration: Configuration for the sync operation.
+    :type rapid_pro_to_engagement_db_configuration: src.rapid_pro_to_engagement_db.configuration.RapidProToEngagementDBConfiguration
+    :param engagement_db: Engagement database to sync to.
+    :type engagement_db: engagement_database.EngagementDatabase
+    :param uuid_table: UUID table to use to de-identify contact urns.
+    :type uuid_table: id_infrastructure.firestore_uuid_table.FirestoreUuidTable
+    """
+    # This implementation is WIP. It shows how we can non-incrementally synchronise a workspace to the database.
+    # To enter production, we still need the following:
+    # TODO: Support incremental update of runs and contacts.
+    # TODO: Handle deleted contacts.
+    # TODO: Handle contacts that have runs but haven't been fetched locally yet.
+    # TODO: Optimise fetching fields from the same flows, so we don't have to download the same runs multiple times.
+
+    rapid_pro = init_rapid_pro_client(google_cloud_credentials_file_path,
+                                      rapid_pro_to_engagement_db_configuration.domain,
+                                      rapid_pro_to_engagement_db_configuration.token_file_url)
+    workspace_name = rapid_pro.get_workspace_name()
+
+    # Build a look-up table of Rapid Pro contact uuid -> Rapid Pro contact so we can look up the urn for each run later.
+    contacts = rapid_pro.get_raw_contacts()
+    contacts_lut = {c.uuid: c for c in contacts}
+
+    for flow_config in rapid_pro_to_engagement_db_configuration.flow_result_configurations:
+        # Get the latest runs for this flow.
+        flow_id = rapid_pro.get_flow_id(flow_config.flow_name)
+        runs = rapid_pro.get_raw_runs(flow_id)
+
+        for i, run in enumerate(runs):
+            log.debug(f"Processing run {i + 1}/{len(runs)}, id {run.id}...")
+
+            # Get the relevant result from this run, if it exists.
+            rapid_pro_result = run.values.get(flow_config.flow_result_field)
+            if rapid_pro_result is None:
+                log.debug("No relevant run result")
+                continue
+
+            # De-identify the contact's urn.
+            # Note (1) this de-identifies the full urn, so we can handle multiple channel types simultaneously.
+            #      (2) we only get this far if there was a valid run result, which ensures we only add de-identification
+            #          entries for participants who messaged us.
+            contact_urn = contacts_lut[run.contact.uuid].urns[0]
+            if contact_urn.startswith("tel:"):
+                assert contact_urn.startswith("tel:+")
```

Comment:
This is not usually the case and not quite sure how we can resolve this because if we normalize while adding to db when we add normalised advert contacts back to textit, it  treats them as two different contacts. - probably build a case and reach out to them.

Reply:
Following offline discussion this only impacts golis via shaqodoon, so I'll add a todo here but leave this fail-safe for now. We'll address fully once we apply this to a project that has this problem